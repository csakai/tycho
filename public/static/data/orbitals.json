[{
	"name": "Earth",
	"albedo": 0.367,
	"sidereal": 0.99726968,
	"radius": 6371,
	"markdown": "earth.md",
	"mass": 5.97237e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 23.4392811,
	"inclination": 7.155,
	"atmosphere": 35260,
	"satellites": [{
		"name": "Moon",
		"albedo": 0.136,
		"sidereal": 0.99726968,
		"radius": 1737.1,
		"markdown": "moon.md",
		"mass": 5.97237e+24,
		"centralMass": 5.97237e+24,
		"axialTilt": 6.687,
		"inclination": 5.145,
		"atmosphere": 10066278,
		"ephemeris": {
			"nasaOrbitalCode": "301",
			"nasaBarycenterCode": "3",
			"revolutionOrder": "d"
		},
		"argumentOfPeriapsis": 282.5992409518417,
		"eccentricity": 0.03405532141489297,
		"longitudeOfAscendingNode": 142.3595722910625,
		"semimajor": 376126.5103019063,
		"semiminor": 375908.33786092995,
		"periapses": {
			"last": 1509927058416.626,
			"next": 1512377208118.1555
		},
		"id": "moon"
	}],
	"ephemeris": {
		"nasaOrbitalCode": "3",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 114.20763,
	"eccentricity": 0.01355360061112275,
	"longitudeOfAscendingNode": 348.73936,
	"semimajor": 150759239.9570005,
	"semiminor": 150745392.07806867,
	"periapses": {
		"last": 1136430000000,
		"next": 1167987600000
	},
	"id": "earth"
},{
	"name": "Venus",
	"albedo": 0.689,
	"sidereal": -243.025,
	"radius": 6051.8,
	"markdown": "venus.md",
	"mass": 4.8675e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 2.64,
	"inclination": 3.86,
	"atmosphere": 16052641,
	"ephemeris": {
		"nasaOrbitalCode": "2",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 55.186,
	"eccentricity": 0.0067,
	"longitudeOfAscendingNode": 76.678,
	"semimajor": 108208179,
	"semiminor": 107477094,
	"periapses": {
		"last": 1429348560000,
		"next": 1448787300000
	},
	"id": "venus"
}]