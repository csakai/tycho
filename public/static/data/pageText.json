{"planetScale":"Planet Scale: × ","speedScale":"Speed Scale: × 10^","settings":"Settings"}
