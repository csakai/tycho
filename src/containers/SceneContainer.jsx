import TWEEN from 'tween.js';
import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import React3 from 'react-three-renderer';
import Scene from '../components/Scene';
import * as AnimationActions from '../actions/AnimationActions';
import * as UIControlsActions from '../actions/UIControlsActions';
import * as TourActions from '../actions/TourActions';
import * as LabelActions from '../actions/LabelActions';
import ReduxService from '../services/ReduxService';
import CameraContainer from './CameraContainer';
import DomEvents from '../utils/DomEvents';

export class SceneContainer extends React.Component {

  static propTypes = {
    orbitalData: PropTypes.array.isRequired,
    onAnimate: PropTypes.func.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    time: PropTypes.number
  }

  /**
   * Animation frame update method.
   */
  onAnimate = () => {
    this.props.onAnimate();
    this.refs.camera.update();
    // this.forceUpdate();
    TWEEN.update();
  }

  /**
   * Maps the mousewheel event to the controls zoom function.
   */
  changeZoom = (ev) => {
    this.refs.camera.controls.wheelZoom(ev, this.props.action.changeZoom);
  }

  /**
   * Starts auto-rotation of the camera.
   */
  startAutoRotate = () => {
    this.props.action.setCameraOrbit(true);
    this.props.action.setUIControls(false);
  }

  /**
   * Stops auto-rotation of the camera.
   */
  stopAutoRotate = () => {
    this.props.action.setCameraOrbit(false);
    this.props.action.setUIControls(true);
  }

  /**
   * Sets the renderer DOM element.
   */
  setDomElement = (domElement) => {
    this.domElement = domElement;
  }

  /**
   * Renders a new Scene.
   *
   * @param {Object} refs
   * @param {Camera} refs.camera - THREE.Camera component instance
   * @returns {Scene} instance of scene component
   */
  renderScene = ({camera}) => {
    return (
      <Scene
        time={this.props.time}
        orbitalData={this.props.orbitalData}
        scale={this.props.scale}
        action={this.props.action}
        children={this.props.children}
        highlightedOrbital={this.props.highlightedOrbital}
        cameraMatrix={camera.position.clone()}
        camera={camera}
        domEvents={DomEvents(camera)}
      />
    );
  }

  render() {
    const {width, height} = this.props;
    const {camera} = this.refs;

    // TODO: move parent div to separate container named "SyntheticEventsContainer"
    return (
      <div
        onWheel={this.changeZoom}
        onTouchStart={this.stopAutoRotate}
        onMouseDown={this.stopAutoRotate}>
        <React3
          onAnimate={this.onAnimate}
          mainCamera="camera"
          width={width}
          height={height}
          antialias={true}
          alpha={true}
          canvasRef={this.setDomElement}>
          <scene ref="scene">
            <CameraContainer
              ratio={width / height}
              targetName={this.props.targetName}
              action={this.props.action}
              speed={this.props.speed}
              scale={this.props.scale}
              scene={this.refs.scene}
              zoom={this.props.zoom}
              isAutoOrbitEnabled={this.props.isAutoOrbitEnabled}
              orbitalData={this.props.orbitalData}
              domElement={this.domElement}
              ref="camera"
            />
            {camera && this.renderScene(camera.refs)}
          </scene>
        </React3>
      </div>
    );
  }
}

export default connect(
  ReduxService.mapStateToProps(
    'uiControls.zoom',
    'uiControls.scale',
    'uiControls.speed',
    'label.targetName',
    'label.highlightedOrbital',
    'tour.isAutoOrbitEnabled',
    'animation.time',
    'data.orbitalData'
  ),
  ReduxService.mapDispatchToProps(
    UIControlsActions,
    TourActions,
    AnimationActions,
    LabelActions
  )
)(SceneContainer);