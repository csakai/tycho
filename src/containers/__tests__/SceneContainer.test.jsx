import React from 'react';
import {CubeTexture} from 'three';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {SceneContainer} from '../SceneContainer';
import data from './__fixtures__/orbitals.json';

describe('Scene Container', () => {
    let component, sceneContainer;

    beforeEach(() => {
        component = shallow(
            <SceneContainer
                orbitalData={data}
                action={{setPosition: jest.fn()}}
                onAnimate={jest.fn()}
                width={500}
                height={300}
                time={1}
            />
        );

        sceneContainer = component.instance();
    });

    describe('componentDidMount()', () => {
        beforeEach(() => {
            sceneContainer.renderSkybox = jest.fn();
        });

        it('should force update once', () => {
            const spy = jest.spyOn(sceneContainer, 'forceUpdate');

            sceneContainer.componentDidMount();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should render the skybox', () => {
            const spy = jest.spyOn(sceneContainer, 'renderSkybox');

            sceneContainer.componentDidMount();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });
    });

    describe('onAnimate()', () => {
        beforeEach(() => {
            sceneContainer.props = {onAnimate: jest.fn()};
            sceneContainer.refs = {camera};
        });

        it('should call props.onAnimate()', () => {
            const onAnimateSpy = jest.spyOn(sceneContainer.props, 'onAnimate');

            sceneContainer.onAnimate();

            expect(onAnimateSpy).toHaveBeenCalled();
            expect(onAnimateSpy).toHaveBeenCalledTimes(1);
        });

        it('should update the camera component', () => {
            const spy = jest.spyOn(sceneContainer.refs.camera, 'update');

            sceneContainer.onAnimate();

            expect(spy).toHaveBeenCalled();
        });
    });

    describe('changeZoom()', () => {
        it('should call controls.wheelZoom()', () => {
            sceneContainer.refs = {camera};
            sceneContainer.props = {action: {}};
            const spy = jest.spyOn(sceneContainer.refs.camera.controls, 'wheelZoom');

            sceneContainer.changeZoom();

            expect(spy).toHaveBeenCalled();
        });
    });

    describe('setDomElement()', () => {
        it('should set `domElement` to the value of the param passed', () => {
            const elem = <canvas />;
            sceneContainer.setDomElement(elem);

            expect(sceneContainer).toHaveProperty('domElement');
            expect(sceneContainer.domElement).toEqual(elem);
        });
    });

    describe('renderSkybox()', () => {
        it('should set the `background` property of the scene to a cube texture', () => {
            sceneContainer.refs = {
                scene: {
                    background: null
                }
            };
            sceneContainer.renderSkybox();

            expect(sceneContainer.refs.scene).toHaveProperty('background');
            expect(sceneContainer.refs.scene.background).toBeInstanceOf(CubeTexture);
        });
    });

    describe('render()', () => {
        it('should render the scene container without the camera', () => {
            expect(toJson(component)).toMatchSnapshot();
        });

        it('should render the scene container with the camera', () => {
            sceneContainer.refs = {camera};
            sceneContainer.render();
            expect(toJson(component)).toMatchSnapshot();
        });
    });
});

const camera = {
    controls: {
        wheelZoom: jest.fn()
    },
    update: jest.fn(),
    refs: {
        camera: {
            position: {
                clone: jest.fn()
            }
        }
    }
};
