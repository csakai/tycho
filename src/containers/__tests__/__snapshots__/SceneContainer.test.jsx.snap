// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene Container render() should render the scene container successfully 1`] = `
<div
  onWheel={[Function]}
>
  <React3
    alpha={true}
    antialias={true}
    canvasRef={[Function]}
    context="3d"
    height={300}
    mainCamera="camera"
    onAnimate={[Function]}
    width={500}
  >
    <scene>
      <group
        position={
          Object {
            "x": 1,
            "y": 1,
            "z": 1,
          }
        }
      >
        <perspectiveCamera
          aspect={1.6666666666666667}
          far={10000}
          fov={50}
          name="camera"
          near={1}
          position={
            Vector3 {
              "x": 300,
              "y": 300,
              "z": 300,
            }
          }
        />
      </group>
      <Scene
        cameraMatrix={null}
        orbitalData={
          Array [
            Object {
              "GM": 6836529,
              "arcRotate": 45.0411,
              "argPeriapsis": 114.20763,
              "atmosphereColor": 8421504,
              "axialTilt": 23.26,
              "eccentricity": 0.01671123,
              "id": "dummyOuter",
              "inclination": 1.57869,
              "longAscNode": 348.73936,
              "name": "Dummy Outer",
              "periapses": Object {
                "last": 1136430000,
                "next": 1167987600,
              },
              "radius": 637800,
              "ring": Object {
                "innerRadius": 6630,
                "outerRadius": 950000,
              },
              "semimajor": 249598261,
              "semiminor": 249556483,
            },
            Object {
              "GM": 6836529,
              "arcRotate": 45.0411,
              "argPeriapsis": 114.20763,
              "atmosphereColor": 8421504,
              "axialTilt": 23.26,
              "eccentricity": 0.01671123,
              "id": "dummyParent",
              "inclination": 15.57869,
              "longAscNode": 348.73936,
              "name": "Dummy Planet",
              "periapses": Object {
                "last": 1156430000,
                "next": 1167987600,
              },
              "radius": 637800,
              "ring": Object {
                "innerRadius": 6630,
                "outerRadius": 950000,
              },
              "satellites": Array [
                Object {
                  "GM": 6836529,
                  "arcRotate": 15.0411,
                  "argPeriapsis": 114.20763,
                  "atmosphereColor": 16711680,
                  "axialTilt": 23.26,
                  "eccentricity": 0.01671123,
                  "id": "dummyChild",
                  "inclination": 1.57869,
                  "longAscNode": 348.73936,
                  "name": "Dummy Satellite",
                  "periapses": Object {
                    "last": 1166430000,
                    "next": 1167987600,
                  },
                  "radius": 637800,
                  "semimajor": 149598261,
                  "semiminor": 149556483,
                },
              ],
              "semimajor": 149598261,
              "semiminor": 149556483,
            },
          ]
        }
        time={1}
        updatePosition={[Function]}
      />
    </scene>
  </React3>
</div>
`;
